auth/threelegged.go

```diff
@@ -0,0 +1,345 @@
+	}
+	if c.RedirectURL != "" {
+		v.Set("redirect_uri", c.RedirectURL)
+	}
+	if c.AuthHandlerOpts != nil &&
+		c.AuthHandlerOpts.PKCEConfig != nil &&
+		c.AuthHandlerOpts.PKCEConfig.Verifier != "" {
+		v.Set(codeVerifierKey, c.AuthHandlerOpts.PKCEConfig.Verifier)
+	}
+	for k := range c.URLParams {
+		v.Set(k, c.URLParams.Get(k))
+	}
+	return fetchToken(ctx, c, v)
+}
+
+// This struct is not safe for concurrent access alone, but the way it is used
+// in this package by wrapping it with a cachedTokenProvider makes it so.
+type tokenProvider3LO struct {
+	opts         *Options3LO
+	client       *http.Client
+	refreshToken string
+}
+
+func (tp *tokenProvider3LO) Token(ctx context.Context) (*Token, error) {
+	if tp.refreshToken == "" {
+		return nil, errors.New("auth: token expired and refresh token is not set")
+	}
+	v := url.Values{
+		"grant_type":    {"refresh_token"},
+		"refresh_token": {tp.refreshToken},
+	}
+	for k := range tp.opts.URLParams {
+		v.Set(k, tp.opts.URLParams.Get(k))
+	}
+
+	tk, rt, err := fetchToken(ctx, tp.opts, v)
+	if err != nil {
+		return nil, err
+	}
+	if tp.refreshToken != rt && rt != "" {
+		tp.refreshToken = rt
+	}
+	return tk, err
+}
+
+type tokenProviderWithHandler struct {
+	opts  *Options3LO
+	state string
+}
+
+func (tp tokenProviderWithHandler) Token(ctx context.Context) (*Token, error) {
+	url := tp.opts.authCodeURL(tp.state, nil)
+	code, state, err := tp.opts.AuthHandlerOpts.Handler(url)
+	if err != nil {
+		return nil, err
+	}
+	if state != tp.state {
+		return nil, errors.New("auth: state mismatch in 3-legged-OAuth flow")
+	}
+	tok, _, err := tp.opts.exchange(ctx, code)
+	return tok, err
+}
+
+// fetchToken returns a Token, refresh token, and/or an error.
+func fetchToken(ctx context.Context, c *Options3LO, v url.Values) (*Token, string, error) {
+	var refreshToken string
+	if c.AuthStyle == StyleUnknown {
+		return nil, refreshToken, fmt.Errorf("auth: missing required field AuthStyle")
+	}
+	if c.AuthStyle == StyleInParams {
+		if c.ClientID != "" {
+			v.Set("client_id", c.ClientID)
+		}
+		if c.ClientSecret != "" {
+			v.Set("client_secret", c.ClientSecret)
+		}
+	}
+	req, err := http.NewRequest("POST", c.TokenURL, strings.NewReader(v.Encode()))
+	if err != nil {
+		return nil, refreshToken, err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	if c.AuthStyle == StyleInHeader {
+		req.SetBasicAuth(url.QueryEscape(c.ClientID), url.QueryEscape(c.ClientSecret))
+	}
+
+	// Make request
+	r, err := c.client().Do(req.WithContext(ctx))
+	if err != nil {
+		return nil, refreshToken, err
+	}
+	body, err := io.ReadAll(io.LimitReader(r.Body, 1<<20))
+	r.Body.Close()
+	if err != nil {
+		return nil, refreshToken, fmt.Errorf("auth: cannot fetch token: %w", err)
+	}
+
+	failureStatus := r.StatusCode < 200 || r.StatusCode > 299
+	tokError := &Error{
+		Response: r,
+		Body:     body,
+	}
+
+	var token *Token
+	content, _, _ := mime.ParseMediaType(r.Header.Get("Content-Type"))
```

Comment:
Just asking: Intentionally squashing the error from here? If so, I'm guessing it's because the `default` case is sufficient?